File: medium/1382_balance_binary_search_tree/main.py
# Given the root of a binary search tree, return a balanced binary search tree with the same node values. If there is more than one answer, return any of them.

# A binary search tree is balanced if the depth of the two subtrees of every node never differs by more than 1.

from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def inOrder(self, node, res):
        if node is None:
            return
        
        self.inOrder(node.left, res)
        res.append(node)
        self.inOrder(node.right, res)

    def performBalance(self, list_nodes, left, right):
        if left > right:
            return None
        
        mid = (left + right) // 2
        root = list_nodes[mid]
        
        root.left = self.performBalance(list_nodes, left, mid - 1)
        root.right = self.performBalance(list_nodes, mid + 1, right)

        return root

    def balanceBST(self, root) -> Optional[TreeNode]:
        inorder_list = []
        self.inOrder(root, inorder_list)
        return self.performBalance(inorder_list, 0, len(inorder_list)-1)


def build_simple_tree(values):
    nodes = [TreeNode(v) for v in values if isinstance(v, int)]
    
    for i in range(len(nodes) - 1):
        nodes[i].right = nodes[i+1]
        
    return nodes[0] if nodes else None


def main():

    node_list = [1,"null",2,"null",3,"null",4,"null","null"]
    root = build_simple_tree(node_list)

    sol = Solution()
    root = sol.balanceBST(root)

    
        

File: medium/1382_balance_binary_search_tree/main.cpp
/*
Given the root of a binary search tree, return a balanced binary search tree with the same node values. If there is more than one answer, return any of them.

A binary search tree is balanced if the depth of the two subtrees of every node never differs by more than 1.
*/

#include <vector>

using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
private:
    void inOrder(TreeNode *node, vector<TreeNode *> &listNodes)
    {
        if (node == nullptr)
        {
            return;
        }

        inOrder(node->left, listNodes);
        node->left = nullptr;
        listNodes.push_back(node);
        inOrder(node->right, listNodes);
        node->right = nullptr;
    }

    TreeNode *doBalance(vector<TreeNode *> &listNodes, size_t left, size_t right)
    {
        if (left > right)
            return nullptr;

        size_t mid = (size_t)((left + right) / 2);
        TreeNode *root = listNodes[mid];

        root->left = doBalance(listNodes, left, mid - 1);
        root->right = doBalance(listNodes, mid + 1, right);

        return root;
    }

public:
    TreeNode *balanceBST(TreeNode *root)
    {
        vector<TreeNode *> inOrderList;
        inOrder(root, inOrderList);
        return doBalance(inOrderList, 0, inOrderList.size() - 1);
    }
};